File: acoustic-field-optimizer/examples/foci.rs
/*
 * File: foci.rs
 * Project: examples
 * Created Date: 21/09/2020
 * Author: Shun Suzuki
 * -----
 * Last Modified: 19/11/2020
 * Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
 * -----
 * Copyright (c) 2020 Hapis Lab. All rights reserved.
 *
 */

use acoustic_field_calculator::prelude::*;
use acoustic_field_optimizer::multiple_foci::*;
use acoustic_field_optimizer::*;

use walkdir::WalkDir;

const NUM_TRANS_X: usize = 18;
const NUM_TRANS_Y: usize = 14;
const TRANS_SIZE: Float = 10.18; // mm
const FREQUENCY: Float = 40e3; // Hz
const TEMPERATURE: Float = 300.0; // K

macro_rules! write_image_xy {
    ($filename: tt, $area: ident, $field: ident) => {{
        use image::png::PngEncoder;
        use image::ColorType;
        use scarlet::colormap::ListedColorMap;
        use std::fs::File;

        let bounds = $area.bounds();
        let bb = (bounds.x(), bounds.y());

        let colormap = ListedColorMap::magma();

        let output = File::create($filename).unwrap();
        let max = $field.max_result() as Float;
        let pixels: Vec<_> = $field
            .buffer()
            .chunks_exact(bb.0)
            .rev()
            .flatten()
            .map(|&v| {
                colormap.vals[(v as Float / max * (colormap.vals.len() - 1) as Float) as usize]
            })
            .flat_map(|c| c.iter().map(|x| (x * 255.) as u8).collect::<Vec<_>>())
            .collect();

        let encoder = PngEncoder::new(output);
        encoder
            .encode(&pixels, bb.0 as u32, bb.1 as u32, ColorType::Rgb8)
            .unwrap();
    }};
}

fn main() {
    let array_center = Vector3::new(
        TRANS_SIZE * (NUM_TRANS_X - 1) as Float / 2.0,
        TRANS_SIZE * (NUM_TRANS_Y - 1) as Float / 2.0,
        0.,
    );
    let z = 150.0;
    let focal_pos = array_center + z * Vector3::z();

    let mut system = UniformSystem::new(TEMPERATURE);
    for y in 0..NUM_TRANS_Y {
        for x in 0..NUM_TRANS_X {
            let pos = Vector3::new(TRANS_SIZE * x as Float, TRANS_SIZE * y as Float, 0.);
            system.add_wave_source(T4010A1::new(pos, Vector3::z(), 1.0, 0.0, FREQUENCY));
        }
    }

    let r = 200.0;
    let area = GridAreaBuilder::new()
        .x_range(array_center[0] - r / 2.0, array_center[0] + r / 2.0)
        .y_range(array_center[1] - r / 2.0, array_center[1] + r / 2.0)
        .z_at(z)
        .resolution(1.)
        .generate();
    let mut field = PressureField::new();

    let calculator = CpuCalculator::new();

    let foci = vec![
        focal_pos + Vector3::new(50., 0., 0.),
        focal_pos - Vector3::new(50., 0., 0.),
    ];
    let amps = vec![1., 1.];

    /////////////////  IFFT  /////////////////////////
    // can only use when 2D phased array and 2D target
    for entry in WalkDir::new(".").max_depth(4) {
        let entry = entry.unwrap();
        let path = entry.path().to_str().unwrap();
        if path.contains("star.bmp") {
            let bottom_left = Vector3::new(0., 0., 0.);
            let top_left = Vector3::new(0., TRANS_SIZE * (NUM_TRANS_Y - 1) as Float, 0.);
            let bottom_right = Vector3::new(TRANS_SIZE * (NUM_TRANS_X - 1) as Float, 0., 0.);

            IFFT::new(path, bottom_left, top_left, bottom_right, TRANS_SIZE, z)
                .optimize(&mut system);
            calculator.calculate(&system, &area, &mut field);
            write_image_xy!("ifft_star.png", area, field);
            break;
        }
    }

    ////////////////////////// Long et al. ///////////////////////////////
    // please specify maximum amplitude before
    for source in system.wave_sources_mut() {
        source.set_amp(1.0);
    }
    Long::new(foci.clone(), amps.clone()).optimize(&mut system);
    calculator.calculate(&system, &area, &mut field);
    write_image_xy!("long.png", area, field);

    /////////////////////////////// HORN ///////////////////////////////
    // please specify maximum amplitude before
    for source in system.wave_sources_mut() {
        source.set_amp(1.0);
    }
    Horn::new(foci.clone(), amps.clone()).optimize(&mut system);
    calculator.calculate(&system, &area, &mut field);
    write_image_xy!("horn.png", area, field);

    /////////////////  Naive linear synthesis  /////////////////////////
    // please specify maximum amplitude before
    for source in system.wave_sources_mut() {
        source.set_amp(1.0);
    }
    Naive::new(foci.clone(), amps.clone()).optimize(&mut system);
    calculator.calculate(&system, &area, &mut field);
    write_image_xy!("naive.png", area, field);

    /////////////////  Gerchberg-Saxton  /////////////////////////
    // please specify maximum amplitude before
    for source in system.wave_sources_mut() {
        source.set_amp(1.0);
    }
    GS::new(foci.clone(), amps.clone()).optimize(&mut system);
    calculator.calculate(&system, &area, &mut field);
    write_image_xy!("gs.png", area, field);

    /////////////////  GS-PAT  /////////////////////////
    // please specify maximum amplitude before
    for source in system.wave_sources_mut() {
        source.set_amp(1.0);
    }
    GSPAT::new(foci.clone(), amps.clone()).optimize(&mut system);
    calculator.calculate(&system, &area, &mut field);
    write_image_xy!("gs-pat.png", area, field);

    /////////////////  Gradient Descent  /////////////////////////
    // Gradient Descent optimizer currently does not support amplitudes
    for source in system.wave_sources_mut() {
        source.set_amp(1.0);
    }
    GradientDescent::new(foci.clone(), amps.clone()).optimize(&mut system);
    calculator.calculate(&system, &area, &mut field);
    write_image_xy!("gradient.png", area, field);

    /////////////////  Gauss Newton  /////////////////////////
    // Gauss Newton optimizer currently does not support amplitudes
    for source in system.wave_sources_mut() {
        source.set_amp(1.0);
    }
    GaussNewton::new(foci.clone(), amps.clone()).optimize(&mut system);
    calculator.calculate(&system, &area, &mut field);
    write_image_xy!("gauss-newton.png", area, field);

    /////////////////  Levenberg-Marquardt  /////////////////////////
    // Levenberg-Marquardt optimizer currently does not support amplitudes
    for source in system.wave_sources_mut() {
        source.set_amp(1.0);
    }
    LM::new(foci.clone(), amps.clone()).optimize(&mut system);
    calculator.calculate(&system, &area, &mut field);
    write_image_xy!("lm.png", area, field);

    /////////////////  Acoustic Power Optimization  /////////////////////////
    // please specify maximum amplitude before
    for source in system.wave_sources_mut() {
        source.set_amp(1.0);
    }
    APO::new(foci, amps, 2.0).optimize(&mut system);
    calculator.calculate(&system, &area, &mut field);
    write_image_xy!("apo.png", area, field);
}

File: python/foci.py
'''
File: foci.py
Project: python
Created Date: 22/09/2020
Author: Shun Suzuki
-----
Last Modified: 19/11/2020
Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
-----
Copyright (c) 2020 Hapis Lab. All rights reserved.

'''

import os  # NOQA
import math  # NOQA
import matplotlib.pyplot as plt
import numpy as np
import mpl_toolkits.axes_grid1

from afc import UniformSystem
from afc import SphereWaveSource, T4010A1  # NOQA
from afc import CpuCalculator, GpuCalculator, GridAreaBuilder, PressureFieldBuffer, PowerFieldBuffer  # NOQA
from afc import Optimizer
from afc import Vector3  # NOQA
from afc import plot_helper


def plot(bounds, results, name, ext='png', show=True):
    array = result.reshape(bounds[0], bounds[1])
    DPI = 72
    fig = plt.figure(figsize=(6, 6), dpi=DPI)
    axes = fig.add_subplot(111, aspect='equal')
    heat_map = plot_helper.plot_acoustic_field_2d(axes, array, X_RANGE, Y_RANGE, RESOLUTION, ticks_step=10.0)
    divider = mpl_toolkits.axes_grid1.make_axes_locatable(axes)
    cax = divider.append_axes('right', '5%', pad='3%')
    fig.colorbar(heat_map, cax=cax)
    plt.savefig(f'{name}.{ext}')
    if show:
        plt.title(name)
        plt.show()


if __name__ == '__main__':
    NUM_TRANS_X = 18
    NUM_TRANS_Y = 14
    TRANS_SIZE = 10.18
    FREQUENCY = 40e3
    TEMPERATURE = 300
    Z_DIR = np.array([0, 0, 1])

    array_center = np.array([TRANS_SIZE * (NUM_TRANS_X - 1) / 2.0, TRANS_SIZE * (NUM_TRANS_Y - 1) / 2.0, 0])

    calculator = CpuCalculator()

    system = UniformSystem(TEMPERATURE)
    for y in range(NUM_TRANS_Y):
        for x in range(NUM_TRANS_X):
            pos = np.array([TRANS_SIZE * x, TRANS_SIZE * y, 0.]) - array_center
            source = T4010A1(pos, Z_DIR, 1.0, 0.0, FREQUENCY)
            system.add_wave_source(source)

    z = 150.0

    R = 200.0
    X_RANGE = (-R / 2, R / 2)
    Y_RANGE = (-R / 2, R / 2)
    RESOLUTION = 1.0

    observe_area = GridAreaBuilder()\
        .x_range(X_RANGE)\
        .y_range(Y_RANGE)\
        .z_at(z)\
        .resolution(RESOLUTION)\
        .generate()
    bounds = observe_area.get_size()

    field = PressureFieldBuffer()

    focal_pos = z * Z_DIR
    foci = [focal_pos + np.array([50., 0, 0]), focal_pos - np.array([50., 0, 0])]
    amps = [1.0, 1.0]

    # # SMILE
    # radius = 45.0
    # num = 30
    # foci = []
    # for i in range(num):
    #     theta = 2 * math.pi * i / num
    #     foci.append(focal_pos + radius * np.array([math.cos(theta), math.sin(theta), 0.0]))
    # foci.append(focal_pos + np.array([radius * 0.3, radius * 0.3, 0]))
    # foci.append(focal_pos + np.array([-radius * 0.3, radius * 0.3, 0]))
    # for i in range(1, num // 4):
    #     theta = -math.pi * i / (num // 4)
    #     foci.append(focal_pos + radius * 0.6 * np.array([math.cos(theta), math.sin(theta), 0.0]))
    # amps = 1.0 * np.ones(len(foci))

    # IFFT
    # can only use when 2d phased array and 2d target
    for source in system.get_wave_sources():
        source.amp = 1.0
    path = os.path.join(os.path.dirname(__file__), '../acoustic-field-optimizer/examples/pattern/star.bmp')
    bottom_left = Vector3(np.array([0., 0., 0.]) - array_center)
    top_left = Vector3(np.array([0., TRANS_SIZE * (NUM_TRANS_Y - 1), 0.]) - array_center)
    bottom_right = Vector3(np.array([TRANS_SIZE * (NUM_TRANS_X - 1), 0., 0.]) - array_center)
    Optimizer.ifft(system, path, bottom_left, top_left, bottom_right, TRANS_SIZE, z)
    result = calculator.calculate(system, observe_area, field)
    plot(bounds, result, 'ifft')

    # GS-PAT
    # please specify maximum amplitude before
    for source in system.get_wave_sources():
        source.amp = 1.0
    Optimizer.gs_pat(system, foci, amps)
    result = calculator.calculate(system, observe_area, field)
    plot(bounds, result, 'gs-pat')

    # Gerchberg-Saxton
    # please specify maximum amplitude before
    for source in system.get_wave_sources():
        source.amp = 1.0
    Optimizer.gs(system, foci, amps)
    result = calculator.calculate(system, observe_area, field)
    plot(bounds, result, 'gs')

    # Naive linear synthesis
    # please specify maximum amplitude before
    for source in system.get_wave_sources():
        source.amp = 1.0
    Optimizer.naive(system, foci, amps)
    result = calculator.calculate(system, observe_area, field)
    plot(bounds, result, 'naive')

    # Horn
    # please specify maximum amplitude before
    for source in system.get_wave_sources():
        source.amp = 1.0
    Optimizer.horn(system, foci, amps)
    result = calculator.calculate(system, observe_area, field)
    plot(bounds, result, 'horn')

    # Long
    # please specify maximum amplitude before
    for source in system.get_wave_sources():
        source.amp = 1.0
    Optimizer.long2014(system, foci, amps)
    result = calculator.calculate(system, observe_area, field)
    plot(bounds, result, 'long')

    # Acoustic Power Optimization (with BFGS)
    for source in system.get_wave_sources():
        source.amp = 1.0
    Optimizer.apo(system, foci, amps, 2.0)
    result = calculator.calculate(system, observe_area, field)
    plot(bounds, result, 'apo')

    # Levenberg-Marquardt
    # Levenberg-Marquardt optimizer currently does not support amplitudes
    for source in system.get_wave_sources():
        source.amp = 1.0
    Optimizer.lm(system, foci, amps)
    result = calculator.calculate(system, observe_area, field)
    plot(bounds, result, 'lm')

    # Gauss-Newton
    # Gauss-Newton optimizer currently does not support amplitudes
    for source in system.get_wave_sources():
        source.amp = 1.0
    Optimizer.gauss_newton(system, foci, amps)
    result = calculator.calculate(system, observe_area, field)
    plot(bounds, result, 'gauss_newton')

    # Gradient-Descent
    # Gradient-Descent optimizer currently does not support amplitudes
    for source in system.get_wave_sources():
        source.amp = 1.0
    Optimizer.gradient_descent(system, foci, amps)
    result = calculator.calculate(system, observe_area, field)
    plot(bounds, result, 'gradient_descent')
